### DISCUSSION TITLE
How to use metadata inside client side?

### DISCUSSION THREAD

**Discussion by dwissaaj** [vercel/next.js#54053] [Help] [answered]
### Summary

I am using route handler with prisma and swr to get request. The problem is how we can use generate metadata for a single dynamic page when it only work for server compenent?

```

- usePathname()
- useParams()
- useRouter()

```

All that function up is work only for client which necessary for dynamic page, but generate metada need a server component. Any Idea how to solve this?
[Project Example](https://github.com/dwissaaj/kisetsu.tech/tree/development)
Thanks

**Comment by icyJoseph** [comment]
There's a couple of ways to get around this. 

- Refactor so that the page.tsx file is a server component, which renders a client component with all of the logic you need
- Use a layout.tsx file that exports the meta data
> https://nextjs.org/docs/app/building-your-application/optimizing/metadata#static-metadata
> To define static metadata, export a [Metadata object](https://nextjs.org/docs/app/api-reference/functions/generate-metadata#metadata-object) from a layout.js or static page.js file.

**Reply by dwissaaj** [reply]
Hi Thanks I have tried your solution strange happen. In dynamic page of next return for example `/this%20is%20blogpost` can we change like normal `/this-is-blogpost`

**Reply by icyJoseph** [reply]
Can you link to a page in your repository where this happens? Or share code showing how you are doing it?

**Comment by dwissaaj** [comment]
So every time we using params nextjs decode it using `%20` for every space, because of that too it affect the metadata that's why the title head for seo render  `post/example%20post` for route `/blog/[title]/page.tsx`
_Image_


[Code Repo](https://github.com/dwissaaj/kisetsu.tech/blob/development/src/app/blog/%5Btitle%5D/page.tsx)

**Reply by icyJoseph** [reply]
Then you need to use https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/decodeURIComponent when calculating the title. 

Not sure it is Next.ja necessarily doing that, as much as it is that spaces are reserved characters, and are encoded by the browser/URL API.

**Reply by dwissaaj** [reply]
That's the problem we can't do it in. DecodeURL need happen before rendering. the default url will still use `20%`  . Is this kinda a bug?

**Reply by icyJoseph** [reply, marked as answer]
I am not sure I understand your problem.

```tsx
export async function generateMetadata({
  params: { title },
}: {
  params: { title: string };
}) {
  return {
    title: decodeURIComponent(title),
  };
}

export default async function Page({
  params: { title },
}: {
  params: { title: string };
}) {
  return (
    <>
      <div className={"container mx-auto"}>
        <div className={"bg-neutral-700 p-2 m-2"}>
          <div className={"grid grid-cols-1 md:grid-cols-2 lg:grid-cols-4"}>
            <p>this is {decodeURIComponent(title)}</p>
          </div>
        </div>
      </div>
    </>
  );
}
```

The above works just fine?

**Reply by dwissaaj** [reply]
Yeah it work for the metadata but the route dynamic still use `%20%` for the url I attached image below . I cannot decode the params because its need to happen before the render. There is also similar question in SO but no one answer it.

[Other related question](https://stackoverflow.com/questions/70135681/how-to-add-dash-between-name-of-dynamic-route-nextjs-in-url)

**Reply by icyJoseph** [reply]
That's because you are trying to break the URL standard. I don't understand how is this even related to the original question anyway. And you are being unclear about your procedure. I will not click that SO link.

Do you want to change `/blog/kaki photobook`, to `/blog/kaki-photobook`? And avoid with that, the ugly `%20` encoding automatically done by the browser?

Then I guess you need to redirect space separated URLs to Hyphen separated.

Try something like this?

```tsx
import { redirect } from "next/navigation";

export async function generateMetadata({
  params: { title },
}: {
  params: { title: string };
}) {
  return {
    title: decodeURIComponent(title),
  };
}

const isWhiteSpaceSeparated = (input: string) => {
  return /.{0,}\s+.?/.test(input);
};

const hyphenSeparated = (input: string) => {
  return input.replace(/\s/g, "-");
};

export default async function Page({
  params: { title },
}: {
  params: { title: string };
}) {
  const decodedTitle = decodeURIComponent(title);

  if (isWhiteSpaceSeparated(decodedTitle)) {
    redirect(hyphenSeparated(decodedTitle));
  }

  return (
    <>
      <div className={"container mx-auto"}>
        <div className={"bg-neutral-700 p-2 m-2"}>
          <div className={"grid grid-cols-1 md:grid-cols-2 lg:grid-cols-4"}>
            <p>this is {decodedTitle}</p>
          </div>
        </div>
      </div>
    </>
  );
}
```

You could also do that in middleware or something. 

Like I said, I think your issue has been answered already. If you have other issues open a new discussion.

**Comment by dwissaaj** [comment]
I forgot to say thank you for helping. For now I am gonna use default decode url. Thanks 🔥
